src/bloom/hierarchical.rs

```diff
@@ -0,0 +1,335 @@
+    async fn check_filter<Child, Key>(&self, children: &[Child], item: &Key) -> Result<Option<bool>>
+    where
+        Child: BloomProvider<Key = Key> + Send + Sync,
+        Key: Send + Sync + AsRef<[u8]> + ?Sized,
+    {
+        match self {
+            HierarchicalBloomInner::Vec(filter, values) => {
+                if let Some(filter) = filter {
+                    if let Ok(true) = filter.contains_in_memory(item) {
+                        return Ok(Some(true));
+                    }
+                }
+                let mut falses = false;
+                for filter in values {
+                    match filter.check_filter(children, item).await {
+                        Ok(Some(true)) => return Ok(Some(true)),
+                        Ok(Some(false)) => {}
+                        _ => {
+                            falses = false;
+                        }
+                    }
+                }
+                if falses {
+                    Ok(Some(false))
+                } else {
+                    Ok(None)
+                }
+            }
+            HierarchicalBloomInner::Item(child) => {
+                children.get(*child).unwrap().check_filter(item).await
+            }
+        }
+    }
+
+    #[async_recursion::async_recursion]
+    async fn offload_buffer<Child, Key>(
+        &mut self,
+        children: &mut [Child],
+        needed_memory: usize,
+    ) -> usize
+    where
+        Child: BloomProvider<Key = Key> + Send + Sync,
+        Key: Send + Sync + AsRef<[u8]> + ?Sized,
+    {
+        match self {
+            Self::Vec(_, values) => {
+                let mut freed = 0;
+                for filter in values {
+                    if freed >= needed_memory {
+                        return freed;
+                    }
+                    freed += filter.offload_buffer(children, needed_memory).await;
+                }
+                freed
+            }
+            Self::Item(child) => {
+                children
+                    .get_mut(*child)
+                    .unwrap()
+                    .offload_buffer(needed_memory)
+                    .await
+            }
+        }
+    }
+
+    fn merge_filters(dest: &mut Option<Bloom>, source: &Option<Bloom>) {
+        match source {
+            Some(source_filter) => {
+                let res = match dest {
+                    Some(dest_filter) => dest_filter.checked_add_assign(source_filter),
+                    _ => None,
+                };
+                if res.is_none() {
+                    *dest = None;
+                }
+            }
+            None => *dest = None,
+        }
+    }
+
+    fn push_group(&mut self, value: Self) {
+        match self {
+            Self::Vec(filter, values) => {
+                match &value {
+                    Self::Vec(child_filter, _) => {
+                        Self::merge_filters(filter, child_filter);
+                    }
+                    _ => unreachable!(),
```

Comment:
I think you should mention `Item` variant explicitly everywhere, otherwise if any other variant will be added in the future no compiler warning will be issued